projects/js-packages/charts/src/providers/chart-context/global-charts-provider.tsx

```diff
@@ -9,27 +9,24 @@ export const GlobalChartsContext = createContext< GlobalChartsContextValue | nul
 
 export interface GlobalChartsProviderProps {
 	children: ReactNode;
-	/** Optional theme override. Considered static for provider lifecycle. */
 	theme?: Partial< ChartTheme >;
 }
 
-export const GlobalChartsProvider: FC< GlobalChartsProviderProps > = ( {
-	children,
-	theme = {},
-} ) => {
+export const GlobalChartsProvider: FC< GlobalChartsProviderProps > = ( { children, theme } ) => {
 	const [ charts, setCharts ] = useState< Map< string, ChartRegistration > >( () => new Map() );
 
-	const providerTheme: CompleteChartTheme = useMemo(
-		() => mergeThemes( defaultTheme, theme ),
-		[ theme ]
-	);
+	const providerTheme: CompleteChartTheme = useMemo( () => {
+		return theme ? mergeThemes( defaultTheme, theme ) : defaultTheme;
+	}, [ theme ] );
 
-	// Stable group -> color mapping for this provider lifecycle
-	const groupToColorMapRef = useRef< Map< string, string > >( new Map() );
+	const [ groupToColorMap, setGroupToColorMap ] = useState< Map< string, string > >(
+		() => new Map()
+	);
 
-	// Reset group color mappings when theme changes
+	// Reset group color mappings when theme colors change
 	useEffect( () => {
-		groupToColorMapRef.current = new Map();
+		// Create a completely new Map instance to trigger dependencies, e.g. useChartLegendItems
+		setGroupToColorMap( new Map() );
 	}, [ providerTheme.colors ] );
```

Comment:
This is the fix for [CHARTS-98: Charts: Improve LegendItems data flow](https://linear.app/a8c/issue/CHARTS-98/charts-improve-legenditems-data-flow); making the color map stateful means that chart legend items are updated after this changes, not before, as was the case when they were getting out of sync.

We're now able to remove the color resolution on render workaround.

Reply:
Love it!